zip-0315.rst

```diff
@@ -1,9 +1,620 @@
+desired transactions to occur, and exposing any privacy implications to the user so
+that they have enough information to assess the consequences of their economic actions.
+
+This ZIP covers best practices for:
+
+* what information to display to the user about transactions and balances;
+* how to handle interactions between the ZIP 32 key tree and Unified Addresses;
+* when to use external or internal keys/addresses;
+* sharing addresses and viewing keys;
+* sending and receiving funds (including fee selection);
+* migrating funds between pools.
+
+
+Requirements
+============
+
+User consent
+------------
+
+A wallet typically reveals some information in the process of creating a transaction.
+Which information is revealed depends on the configured wallet privacy policy.
+The guiding principle of these requirements is that users must explicitly consent
+to each instance of information being revealed by the wallet in a transaction.
+
+A user may give blanket consent to reveal a particular kind of information, and
+must also be able to change the configured wallet privacy policy to avoid the
+wallet creating new information leaks of a given type.
+
+The specifications below describe some situations in which blanket consent may be
+inappropriate.
+
+Some varieties of consent may not be revocable, for example if a user chooses to
+share some of their keys.
+
+
+Prompt accessibility of funds
+-----------------------------
+
+Wallets need to take account of two concerns:
+
+* enabling funds to be spent as quickly as possible to reduce latency;
+* waiting for long enough before spending TXOs to ensure that the
+  confirmed-spendable balance is not overestimated, and so can be trusted
+  by the user.
+
+To enable this we define two kinds of TXOs:
+
+* An untrusted TXO is a TXO received from a party that may try to double-spend.
+* A trusted TXO is a TXO received from a party where the wallet trusts a
+  double-spend not to occur, e.g. TXOs created by the wallet's internal TXO
+  handling.
+
+Wallets can then require that untrusted TXOs need more confirmations before
+they become spendable than trusted TXOs. This provides an improved trade-off
+between latency on the one hand, and reliability and safety on the other.
+
+
+Specification
+=============
+
+Long-term storage of funds
+--------------------------
+
+It is RECOMMENDED that wallets only hold funds as shielded in the long-term;
+that is, if a wallet supports receiving transparent funds (or supports
+importing a seed from another wallet that may have done so), then it SHOULD
+either auto-shield or opportunistically shield such funds by default.
+
+Following this recommendation improves both collective privacy and the user's
+individual privacy, by maximizing the size of the note anonymity set over time.
+
+The remainder of this specification assumes a wallet that follows the above
+recommendation, except where explicitly noted.
+
+A wallet MAY allow users to disable auto-shielding.
```

Comment:
Shouldn’t a user have to _enable_ auto-(shielding|migrating)?

Since this involves a dedicated tx, it will always incur a fee, which seems like something a user needs to approve.

In the worst case (which would be mitigated by ZIP 317), an attacker could drain wallets that have naïve auto-shielding enabled – the attacker sends a single tx with 1 zat transparent outputs (likely many outputs at different addresses, to make it economically viable for them). Recipients’ wallets auto-shield the 1 zat at a cost of 1,000 (to 5,000) zat.

So, we should add that wallets MUST only create economically-viable auto-(shielding|migrating) txs. Currently this can mean “shield a lot of TXOs at once”, but with 317 it would change to something like “always do at least 2 TXOs, and each one must be at least _n_ * (total_fee / |inputs|)” (where maybe _n_ = 1.5 or something configurable).

Opportunistic (shielding|migration) should also be specified to only happen when it won’t increase the fee. Currently, that is always the case, but again, 317 (or some future change) could alter that).

Reply:
Auto-shielding should never apply to inputs where the cost of shielding them will exceed their economic value.

Reply:
(For reference see https://github.com/zcash/librustzcash/pull/694/commits/847ba49761fdfec1213a6d6417db6165c549241e)

Reply:
One possibility would be that, in the case that the Zcash Sustainability Fund proposal is adopted/implemented, is that one could have a "dump dust to the sustainability fund" exemption to the fee rule for block inclusion. That being said, I also think that dust never being spent is fine, because that's equivalent to burning, which reduces the monetary supply. If ZEC some day becomes much more valuable, then what is currently "dust" could then have economic value, and by that time hopefully one of the scalability proposals will have it such that the fees are low enough for it to be worth actually spending the dust on goods and services.

Reply:
One other thing here is that since spends must have their proofs checked, increasing the number of spends to spend this dust does impose a cost on the ecosystem, so the spending entity needs to pay a fee to compensate the ecosystem for validating their spend, so from the perspective of the user wanting to "donate" their dust it's a bit upside-down; they might as well just let the dust sit there and have the effects of the resultant deflation of the monetary base go to benefit the whole ecosystem rather than "recycle" the numerical value of the funds.

Reply:
You can't have exceptions to requiring the ZIP 317 fee for dust inputs, because dust inputs are as usable for the kind of DoS attacks that motivated ZIP 317 as any other input: they require verifying a spend auth signature and potentially a zk proof.

Reply:
In any case, considerations about *how* to auto-shield and/or opportunistically shield should probably go in a separate section to this "Long-term storage of funds" section. This line is correct as written, but should probably be generalized:
```suggestion
A wallet MAY allow users to disable auto-shielding, auto-migration,
opportunistic shielding, and/or opportunistic migration. If it does so,
this need not be via four independent settings.
```